### PR TITLE
Add support for reading pre-existing Federated BigQuery Data Sources in

### LINKED ISSUE #32: Support for querying federated tables in BigQuery
In this mode, the files may already be available in GCS, in which case the connector needs to query directly from GCS using the table metadata from BigQuery instead of issuing an extra export.

### PR DESCRIPTION
**Pull request by pmkc**
Google Cloud Stoage.

Fixes #32 

I haven't looked at integration tests yet. Doing that now.

### CODE REVIEW

File: bigquery/src/main/java/com/google/cloud/hadoop/io/bigquery/AbstractBigQueryInputFormat.java [hunk 1]
```diff
@@ -1,19 +1,20 @@
 package com.google.cloud.hadoop.io.bigquery;
 
+import com.google.api.client.repackaged.com.google.common.annotations.VisibleForTesting;
```
Comment (AngusDavis) [inline bigquery/src/main/java/com/google/cloud/hadoop/io/bigquery/AbstractBigQueryInputFormat.java]: com.google.common.annotations.VisibleForTesting.

File: bigquery/src/main/java/com/google/cloud/hadoop/io/bigquery/AbstractBigQueryInputFormat.java [hunk 2]
```diff
@@ -40,6 +46,11 @@
    */
   public static final String INPUT_FORMAT_CLASS_KEY = "mapreduce.inputformat.class";
 
+  public static final String EXTERNAL_TABLE_TYPE = "EXTERNAL";
```
Comment (AngusDavis) [inline bigquery/src/main/java/com/google/cloud/hadoop/io/bigquery/AbstractBigQueryInputFormat.java]: Java doc

File: bigquery/src/main/java/com/google/cloud/hadoop/io/bigquery/AbstractBigQueryInputFormat.java [hunk 3]
```diff
@@ -178,12 +190,27 @@ private static Export constructExport(
...
         BigQueryConfiguration.DELETE_INTERMEDIATE_TABLE_KEY,
         BigQueryConfiguration.DELETE_INTERMEDIATE_TABLE_DEFAULT);
     String query = configuration.get(BigQueryConfiguration.INPUT_QUERY_KEY);
 
+
+    if (EXTERNAL_TABLE_TYPE.equals(table.getType())) {
+      if (Strings.isNullOrEmpty(query)) {
+        LOG.info("Table is already external, so skipping export");
+        // Otherwise getSplits gets confused.
+        setEnableShardedExport(configuration, false);
```
Comment (AngusDavis) [inline bigquery/src/main/java/com/google/cloud/hadoop/io/bigquery/AbstractBigQueryInputFormat.java]: Does the above enableShardedExport var need to be set as well?

File: bigquery/src/main/java/com/google/cloud/hadoop/io/bigquery/ExportFileFormat.java [hunk 4]
```diff
@@ -28,7 +28,7 @@ private ExportFileFormat(String filePattern, String formatIdentifier) {
   /**
    * Get the file pattern to use when exporting
    */
-  public String getFilePattern() {
+public String getFilePattern() {
```
Comment (AngusDavis) [inline bigquery/src/main/java/com/google/cloud/hadoop/io/bigquery/ExportFileFormat.java]: Revert

File: bigquery/src/main/java/com/google/cloud/hadoop/io/bigquery/AbstractBigQueryInputFormat.java (hunk 2 above)
Comment (pmkc) [inline bigquery/src/main/java/com/google/cloud/hadoop/io/bigquery/AbstractBigQueryInputFormat.java]: Done

File: bigquery/src/main/java/com/google/cloud/hadoop/io/bigquery/ExportFileFormat.java (hunk 4 above)
Comment (pmkc) [inline bigquery/src/main/java/com/google/cloud/hadoop/io/bigquery/ExportFileFormat.java]: Done

File: bigquery/src/main/java/com/google/cloud/hadoop/io/bigquery/AbstractBigQueryInputFormat.java (hunk 3 above)
Comment (pmkc) [inline bigquery/src/main/java/com/google/cloud/hadoop/io/bigquery/AbstractBigQueryInputFormat.java]: No, but I moved it anyways.

**Comment by pmkc** [comment]
PTAL manually ran integration tests, which passed.

**Comment by AngusDavis** [comment]
LGTM. <Insert standard squash before submitting spiel here>

**Pull request closed without merging**